gunrock/app/mf/mf_test.cuh

```diff
@@ -7,324 +7,605 @@
+void DisplaySolution(GraphT graph, ValueT* h_flow, VertexT* reverse, 
+	VertexT sink, VertexT nodes)
 {
-    if (length > 40)
-        length = 40;
+    typedef typename GraphT::CsrT CsrT;
+    typedef typename GraphT::SizeT SizeT;
+    ValueT flow_incoming_sink = 0;
+    SizeT e_start = graph.CsrT::GetNeighborListOffset(sink);
+    SizeT num_neighbors = graph.CsrT::GetNeighborListLength(sink);
+    SizeT e_end = e_start + num_neighbors;
+    for (auto e = e_start; e < e_end; ++e)
+    {
+	ValueT flow = h_flow[reverse[e]];
+	if (util::isValid(flow))
+	    flow_incoming_sink += flow;
+    }
 
-    util::PrintMsg("[", true, false);
-    for (SizeT i = 0; i < length; ++i)
+    util::PrintMsg("The maximum amount of flow that is feasible to reach \
+	    from source to sink is " + std::to_string(flow_incoming_sink), 
+	    true, false);
+}
+/**
+ * @brief For given vertex v, find neighbor which the smallest height.  
+ *
+ * @tparam ValueT	Type of capacity/flow/excess
+ * @tparam VertxeT	Type of vertex
+ * @tparam GraphT	Type of graph
+ * @param[in] graph	Graph
+ * @param[in] x		Index of vertex 
+ * @param[in] height	Function of height on nodes 
+ * @param[in] capacity	Function of capacity on edges
+ * @param[in] flow	Function of flow on edges
+ *
+ * return Index the lowest neighbor of vertex x
+ */
+template<typename ValueT, typename VertexT, typename GraphT>
+VertexT find_lowest(GraphT graph, VertexT x, VertexT* height, ValueT* flow, 
+	VertexT source){
+    typedef typename GraphT::CsrT CsrT;
+    auto e_start = graph.CsrT::GetNeighborListOffset(x);
+    auto num_neighbors = graph.CsrT::GetNeighborListLength(x);
+    auto e_end = e_start + num_neighbors;
+    VertexT lowest;
+    VertexT lowest_id = -1; //graph.edges is unreachable value
+    for (auto e = e_start; e < e_end; ++e)
     {
-        util::PrintMsg(std::to_string(i) + ":"
-            + std::to_string(array[i]) + " ", true, false);
+	if (graph.CsrT::edge_values[e] - flow[e] > (ValueT)0){
+	    auto y = graph.CsrT::GetEdgeDest(e);
+	    if (lowest_id == -1 || height[y] < lowest){
+		lowest = height[y];
+		lowest_id = e;
+	    }
+	}
+    }
+    return lowest_id;
+}
+/**
+  * @brief Relabel: increases height of given vertex
+  *
+  * @tparam ValueT	Type of capacity/flow/excess
+  * @tparam VertxeT	Type of vertex
+  * @tparam GraphT	Type of graph
+  * @param[in] graph	Graph
+  * @param[in] x	Index of vertex
+  * @param[in] height	Function of height on nodes
+  * @param[in] capacity Function of capacity on edges
+  * @param[in] flow	Function of flow on edges
+  *
+  * return True if something changed, false otherwise
+  */
+template<typename ValueT, typename VertexT, typename GraphT>
+bool relabel(GraphT graph, VertexT x, VertexT* height, ValueT* flow, 
+	VertexT source){
+    typedef typename GraphT::CsrT CsrT;
+    auto e = find_lowest(graph, x, height, flow, source);
+    // graph.edges is unreachable value = there is no valid neighbour
+    if (e != -1){
```

Comment:
Could you possibly use `utill::isValid(e)` for this? Because you are comparing `auto`, which could be any type to `-1`, which may result in integer conversion resulting in a change of sign.

Reply:
I change it.